src/md_lammps.c

```diff
@@ -473,4 +474,527 @@ static md_molecule_loader_i lammps_api = {
+	ASSERT(header);
+
+	*header = dataPtr->header;
+	return true;
+}
+
+// This is lowlevel cruft for enabling parallel loading and decoding of frames
+// Returns size in bytes of frame, frame_data_ptr is optional and is the destination to write the frame data to.
+int64_t lammps_fetch_frame_data(struct md_trajectory_o* inst, int64_t frame_idx, void* frame_data_ptr) {
+	md_lammps_trajectory_t* traj_data = (md_lammps_trajectory_t*)inst;
+	ASSERT(traj_data);
+	ASSERT(traj_data->magic == MD_LAMMPS_TRAJ_MAGIC);
+
+	if (!traj_data->filesize) {
+		MD_LOG_ERROR("File size is zero");
+		return 0;
+	}
+
+	if (!traj_data->file) {
+		MD_LOG_ERROR("File handle is NULL");
+		return 0;
+	}
+
+	if (!traj_data->frame_offsets) {
+		MD_LOG_ERROR("Frame offsets is empty");
+		return 0;
+	}
+
+	if (!(0 <= frame_idx && frame_idx < (int64_t)md_array_size(traj_data->frame_offsets) - 1)) {
+		MD_LOG_ERROR("Frame index is out of range");
+		return 0;
+	}
+
+	const int64_t beg = traj_data->frame_offsets[frame_idx + 0];
+	const int64_t end = traj_data->frame_offsets[frame_idx + 1];
+	const int64_t frame_size = end - beg;
+
+	if (frame_data_ptr) {
+		// Store the index to the frame since this is generally not found within the actual frame data
+		int64_t* ptr = (int64_t*)frame_data_ptr;
+		ptr[0] = frame_idx;
+		ASSERT(traj_data->file);
+		md_mutex_lock(&traj_data->mutex);
+		md_file_seek(traj_data->file, beg, MD_FILE_BEG);
+		const int64_t bytes_read = md_file_read(traj_data->file, &ptr[1], frame_size);
+		md_mutex_unlock(&traj_data->mutex);
+		ASSERT(frame_size == bytes_read);
+	}
+
+	return frame_size;
+}
+
+bool lammps_decode_frame_data(struct md_trajectory_o* inst, const void* frame_data_ptr, int64_t frame_data_size, md_trajectory_frame_header_t* header, float* x, float* y, float* z) {
+	ASSERT(inst);
+	ASSERT(frame_data_ptr);
+	ASSERT(frame_data_size);
+
+	str_t tokens[32];
+	int64_t num_tokens = 0;
+	int64_t num_token_types = 0;
+
+	const int64_t* ptr = frame_data_ptr;
+	int32_t frame_idx = (int32_t)ptr[0];
+
+
+	md_lammps_trajectory_t* traj_data = (md_lammps_trajectory_t*)inst;
+	if (traj_data->magic != MD_LAMMPS_TRAJ_MAGIC) {
+		MD_LOG_ERROR("Error when decoding frame header, lammps magic did not match");
+		return false;
+	}
+	if (frame_idx < 0 || frame_idx >= traj_data->header.num_frames) {
+		MD_LOG_ERROR("Error when decoding frame data, corrupt frame index");
+		return false;
+	}
+
+	str_t str = { .ptr = (const char*)(ptr + 1), .len = frame_data_size };
+	str_t line;
+
+	//Parse the ITEM: ATOMS to see what type if coordinates we are working with
+	str_t coord_type = { 0 };
+	int32_t coord_start = -1;
+	if (str_extract_line(&line, &str)) {
+		if (str_equal_cstr_n(line, "ITEM: ATOMS", 11)) {
+			num_tokens = extract_tokens(tokens, 32, &line);
+			num_token_types = num_tokens - 2;
+			if (num_token_types < 3) {
+				MD_LOG_ERROR("Fewer than 3 token types in ITEM: ATOMS line");
+				return false;
+			}
+			for (int32_t i = 0; i < num_tokens; i++) {
```

Comment:
You 'potentially' read out of bounds here when checking the tokens, you iterate up until < num_tokens, but access [i + 2]

Reply:
Changed this to use `num_token_types = num_tokens - 2` instead in the loop, this should stop any overflow as there is always 2 more tokens after the last index.